sweep/sweeper.go

```diff
@@ -30,13 +73,585 @@ type UtxoSweeperConfig struct {
+	Err error
+	Tx  *wire.MsgTx
 }
 
-// New returns a new UtxoSweeper instance.
+type sweeperInput struct {
+	input      Input
+	resultChan chan Result
+}
+
+type inputState uint8
+
+const (
+	stateNew = iota
+	statePublished
+	stateError
+)
+
+type pendingInput struct {
+	txHash        chainhash.Hash
+	listeners     []chan Result
+	input         Input
+	ntfnRegCancel func()
+	state         inputState
+	addedWeight   int
+}
+
+// inputYield is a helper struct to sort inputs by yield.
+type inputYield struct {
+	input    *pendingInput
+	netYield int64
+}
+
+// New returns a new Sweeper instance.
 func New(cfg *UtxoSweeperConfig) *UtxoSweeper {
+	// Calculate dust limit based on the P2WPKH output script of the sweep
+	// txes.
+	dustLimit := txrules.GetDustThreshold(
+		lnwallet.P2WPKHSize,
+		btcutil.Amount(cfg.Estimator.RelayFeePerKW().FeePerKVByte()))
+
 	return &UtxoSweeper{
-		cfg: cfg,
+		cfg:           cfg,
+		newInputs:     make(chan *sweeperInput),
+		spendChan:     make(chan *chainntnfs.SpendDetail),
+		quit:          make(chan struct{}),
+		pendingInputs: make(map[wire.OutPoint]*pendingInput),
+		dustLimit:     dustLimit,
+	}
+}
+
+// Start starts the process of constructing and publish sweep txes.
+func (s *UtxoSweeper) Start() error {
+
+	if !atomic.CompareAndSwapUint32(&s.started, 0, 1) {
+		return nil
+	}
+
+	log.Debugf("Starting sweeper")
+
+	s.cfg.Ticker.Resume()
+
+	s.wg.Add(1)
+	go s.collector()
+
+	return nil
+}
+
+// Stop stops sweeper from listening to block epochs and constructing sweep
+// txes.
+func (s *UtxoSweeper) Stop() error {
+	if !atomic.CompareAndSwapUint32(&s.stopped, 0, 1) {
+		return nil
+	}
+
+	log.Infof("Sweeper shutting down")
+
+	close(s.quit)
+	s.wg.Wait()
+
+	log.Infof("Sweeper shut down")
+
+	return nil
+}
+
+// SweepInput sweeps inputs back into the wallet. The inputs will be batched and
+// swept after the batch time window ends.
+func (s *UtxoSweeper) SweepInput(input Input) (chan Result, error) {
+	log.Debugf("Sweep request: %v", input.OutPoint())
+
+	sweeperInput := &sweeperInput{
+		input:      input,
+		resultChan: make(chan Result, 1),
+	}
+
+	// Deliver input to main event loop.
+	select {
+	case s.newInputs <- sweeperInput:
+	case <-s.quit:
+		return nil, fmt.Errorf("sweeper shutting down")
+	}
+
+	return sweeperInput.resultChan, nil
+}
+
+// collector is the sweeper main loop. It processes new inputs, spend
+// notifications and counts down to publication of the sweep tx.
+func (s *UtxoSweeper) collector() {
+	defer s.wg.Done()
+	defer s.cfg.Ticker.Stop()
+
+	log.Debug("Collector started")
+
+	for {
+		select {
+		case input := <-s.newInputs:
+			outpoint := *input.input.OutPoint()
+			pendInput, pending := s.pendingInputs[outpoint]
+			if pending {
+				log.Debugf("Already pending input %v received",
+					outpoint)
+
+				// Add additional result channel to signal spend
+				// of this input.
+				pendInput.listeners =
+					append(pendInput.listeners,
+						input.resultChan)
+
+				break
```

Comment:
Usually `continue` is used here.

Reply:
@Roasbeef suggested this too. Will change it here and in future code.